Frameworks/Foundation/_NSInvocation.arm.mm

```diff
@@ -0,0 +1,486 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import "NSInvocationInternal.h"
+
+#import <Starboard/SmartTypes.h>
+#import <objc/runtime.h>
+#import <objc/encoding.h>
+
+#import <cstdint>
+#import <bitset>
+#import <vector>
+#import <algorithm>
+
+constexpr size_t GPR_COUNT = 4;
+constexpr size_t g_gprLength = GPR_COUNT * sizeof(uintptr_t); // 4 GPRs;
+constexpr size_t SPFR_COUNT = 16;
+constexpr size_t g_sfprLength = SPFR_COUNT * sizeof(float); // 16 single-precision float registers
+
+enum _arm_return_type {
+    ARM_NONE,
+    ARM_VFP_S,
+    ARM_VFP_D,
+    ARM_VFP_HOMOGENOUS,
+    ARM_INT,
+    ARM_INT64,
+    ARM_STRUCT,
+};
+
+// uniformTypeFromStructSpecifier attempts to determine the common type for an aggregate.
+// For aggregates of disparate types or unknowable types, this function returns \0.
+// struct type encodings are generally of the form {name=mmm}, where m is a member encoding.
+// This function requires a char** so that it may advance the type encoding in the recursive case.
+//
+// Base case:
+// {CGSize=ff}
+//  ====f====  <- homogenous type is f
+//
+// {Struct=il}
+// --- no homogenous type ---
+//
+// Recursive case:
+// {CGRect={CGSize=ff}{CGPoint=ff}}
+//          ====f====  =====f==== <- recursively resolve homogenous type of sub-structures
+//              |           |
+//              v           v
+//  ==============f============== <- resolve homogenous type for CGRect to f
+static char uniformTypeFromStructSpecifier(const char** type) {
+    char uniformType = '\0';
+    do {
+        ++(*type);
+        if (_C_STRUCT_E == **type) {
+            // premature end: forward-declared struct with unknown members.
+            return '\0';
+        }
+    } while ('=' != **type);
+
+    ++(*type); // skip the =
+
+    while (**type != _C_STRUCT_E) {
+        if (**type == '"') {
+            // Occasionally, struct encodings store field names as in {name="x"f"y"f}. Skip those.
+            do {
+                ++(*type);
+            } while ('"' != **type);
+            ++(*type);
+        }
+        char v = **type;
+        switch (v) {
```

Comment:
No particular reason; almost every other type-switch we do us a switch

---
In reply to: [75930104](https://github.com/Microsoft/WinObjC/pull/903#discussion_r75930104) [](ancestors = 75930104)